splice_machine/src/main/java/com/splicemachine/derby/stream/function/CloneFirstFunction.java

```diff
@@ -0,0 +1,37 @@
+package com.splicemachine.derby.stream.function;
+
+import com.splicemachine.db.iapi.sql.execute.ExecRow;
+import org.apache.spark.api.java.function.Function;
+import org.apache.spark.api.java.function.PairFunction;
+import scala.Tuple2;
+
+import java.util.*;
+
+/**
+ * Created by jleach on 8/30/17.
+ */
+public class CloneFirstFunction implements PairFunction<Tuple2<ExecRow,ExecRow>,ExecRow,ExecRow > {
```

Comment:
Why is this needed? I don't understand very well what it's doing.

Reply:
The way the key by logic works it will never give you a chance to clone.  Basically the internal spark code flow is stick in map if not exists or in map, run reduce function with existing value.   

For example (key,value),

(1,1),
(1,2),
(1,3)
(2,1)
...

The reduce by key starts with ((1,1),(1,2)).  I am already screwed since I needed to clone (1,1) before it got into the map for reduceByKey.  Since I am not cloning all over my value becomes (1,1),(1,1) since they are referring to the same DVD.

Reply:
I see, thanks for the explanation!